add docs for StructuredAuthenticationConfig v1alpha1 (#43397)
kubernetes/website: content/en/docs/reference/access-authn-authz/authentication.md

{{< /mermaid >}}

1.  Login to your identity provider
1.  Your identity provider will provide you with an `access_token`, `id_token` and a `refresh_token`
1.  When using `kubectl`, use your `id_token` with the `--token` flag or add it directly to your `kubeconfig`
1.  `kubectl` sends your `id_token` in a header called Authorization to the API server

> 1.  The API server will make sure the JWT signature is valid by checking against the certificate named in the configuration

Reviewer: ```suggestion
1.  The API server will make sure the JWT signature is valid
```

| `--oidc-ca-file` | The path to the certificate for the CA that signed your identity provider's web certificate.  Defaults to the host's root CAs. | `/etc/kubernetes/ssl/kc-ca.pem` | No |
| `--oidc-signing-algs` | The signing algorithms accepted. Default is "RS256". | `RS512` | No |

##### Using Authentication Configuration

{{< feature-state for_k8s_version="v1.29" state="alpha" >}}

> The API server can be configured to use a JWT authenticator via the `--authentication-config` flag. This flag takes a path to a file containing the `AuthenticationConfiguration`. An example configuration is provided below.

Reviewer: Need to explain that `JWT authenticator` is the generic version of OIDC server config.

> In order to use the feature, the `StructuredAuthenticationConfiguration` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)

Reviewer: ```suggestion
To use this config, the `StructuredAuthenticationConfiguration` [feature](/docs/reference/command-line-tools-reference/feature-gates/)
```

has to be enabled.

Note: When the feature is enabled, setting both `--authentication-config` and any of the `--oidc-*` flags will result in an error. If you want to use the feature, you have to remove the `--oidc-*` flags and use the configuration file instead.

```yaml
---
#
# CAUTION: this is an example configuration.
#          Do not use this for your own cluster!
#
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: AuthenticationConfiguration
# list of authenticators to authenticate Kubernetes users using JWT compliant tokens.
jwt:

> - issuer:

Reviewer: I think you need a few JSON payload examples (I would include full tokens signed by some random key) that would get run through this config so that you can describe what would happen based on those as inputs.

    url: https://example.com # Same as --oidc-issuer-url.
    audiences:
    - my-app # Same as --oidc-client-id.
  # rules applied to validate token claims to authenticate users.
  claimValidationRules:
    # Same as --oidc-required-claim key=value.
  - claim: hd
    requiredValue: example.com
    # Instead of claim and requiredValue, you can use expression to validate the claim.
    # expression is a CEL expression that evaluates to a boolean.
  - expression: 'claims.hd == "example.com"'

>     # Message customizes the error message when the validation fails.

Reviewer: ```suggestion
    # Message customizes the error message seen in the API server logs when the validation fails.
```

    message: the hd claim must be set to example.com
  - expression: 'claims.exp - claims.nbf <= 86400'
    message: total token lifetime must not exceed 24 hours
  claimMappings:
    # username represents an option for the username attribute.
    # This is the only required attribute.
    username:
      # Same as --oidc-username-claim. Mutually exclusive with username.expression.
      claim: "sub"
      # Same as --oidc-username-prefix. Mutually exclusive with username.expression.

>       # if username.claim is set, username.prefix is required.

Reviewer: ```suggestion
      # if username.claim is set, username.prefix is required.
      # Explicitly set it to "" if no prefix is desired.
```

      prefix: ""
      # Mutually exclusive with username.claim and username.prefix.
      # expression is a CEL expression that evaluates to a string.
      expression: 'claims.username + ":external-user"'
    # groups represents an option for the groups attribute.
    groups:
      # Same as --oidc-groups-claim. Mutually exclusive with groups.expression.
      claim: "sub"
      # Same as --oidc-groups-prefix. Mutually exclusive with groups.expression.

>       # if groups.claim is set, groups.prefix is required.

Reviewer: ```suggestion
      # if groups.claim is set, groups.prefix is required.
      # Explicitly set it to "" if no prefix is desired.
```

      prefix: ""

>       # Mutually exclusive with username.claim and groups.prefix.

Reviewer: ```suggestion
      # Mutually exclusive with groups.claim and groups.prefix.
```

      # expression is a CEL expression that evaluates to a string or a list of strings.
      expression: 'claims.roles.split(",")'
    # uid represents an option for the uid attribute.
    uid:
      # Mutually exclusive with uid.expression.
      claim: 'sub'
      # Mutually exclusive with uid.claim
      # expression is a CEL expression that evaluates to a string.
      expression: 'claims.sub'

>     # extra attributes to be added to the UserInfo object. Keys must be domain-prefix path and must be unique.

Reviewer: Need link to schema of user info object.

    extra:
    - key: 'example.com/tenant'
      # valueExpression is a CEL expression that evaluates to a string or a list of strings.
      valueExpression: 'claims.tenant'
  # validation rules applied to the final user object.
  userValidationRules:
    # expression is a CEL expression that evaluates to a boolean.
  - expression: "!user.username.startsWith('system:')"

>     # message customizes the error message when the validation fails.

Reviewer: ```suggestion
    # message customizes the error message seen in the API server logs when the validation fails.
```

    message: 'username cannot used reserved system: prefix'
  - expression: "user.groups.all(group, !group.startsWith('system:'))"
    message: 'groups cannot used reserved system: prefix'
```

* Claim validation rule expression

>   `jwt.claimValidationRules[i].expression` represents the expression which will be evaluated by CEL.

Reviewer: This gets rendered on the same line which does not look correct.

>   CEL expressions have access to the contents of the token claims, organized into `claims` CEL variable.

Reviewer: ```suggestion
  CEL expressions have access to the contents of the token payload, organized into `claims` CEL variable.
```

>   `claims` is a map of claim names to claim values.

Reviewer: ```suggestion
  `claims` is a map of claim names (as strings) to claim values (of any type).
```

* User validation rule expression
  `jwt.userValidationRules[i].expression` represents the expression which will be evaluated by CEL.

>   CEL expressions have access to the contents of userInfo, organized into `user` CEL variable.

Reviewer: ```suggestion
  CEL expressions have access to the contents of `userInfo`, organized into `user` CEL variable.
  `user` is defined as <link to schema>
```

* Claim mapping expression
  `jwt.claimMappings.username.expression`, `jwt.claimMappings.groups.expression`, `jwt.claimMappings.uid.expression`
  `jwt.claimMappings.extra[i].valueExpression` represents the expression which will be evaluated by CEL.
  CEL expressions have access to the contents of the token claims, organized into `claims` CEL variable.
  `claims` is a map of claim names to claim values.

>   To learn more, see the [CEL language specification](https://github.com/google/cel-spec)

Reviewer: I think it would be better to link to the kube specific CEL docs.
